torchmetrics/image/map.py

```diff
@@ -0,0 +1,369 @@
+        return metrics
+
+    def _get_coco_format(self, is_pred: bool = False) -> Dict:
+        """Transforms and returns all cached targets or predictions in COCO format.
+
+        Format is defined at https://cocodataset.org/#format-data
+        """
+
+        images = []
+        annotations = []
+        annotation_id = 1  # has to start with 1, otherwise COCOEval results are wrong
+
+        for i, _ in enumerate(self.groundtruth_boxes):
+            images.append({"id": i})
+
+            if is_pred:
+                boxes = self.detection_boxes[i]
+                classes = self.detection_labels[i]
+                scores = self.detection_scores[i]
+                if boxes.size()[0] != scores.size()[0]:
+                    raise ValueError(
+                        f"Input boxes and scores of sample {i} have a different"
+                        f" length (expected {boxes.size()[0]} scores, got {scores.size()[0]})"
+                    )
+            else:
+                boxes = self.groundtruth_boxes[i]
+                classes = self.groundtruth_labels[i]
+                scores = None
+
+            if boxes.size()[0] != classes.size()[0]:
+                raise ValueError(
+                    f"Input boxes and classes of sample {i} have a"
+                    f" different length (expected {boxes.size()[0]} classes, got {classes.size()[0]})"
+                )
+
+            boxes = boxes.cpu().tolist()
+            classes = classes.cpu().tolist()
+            if scores is not None:
+                scores = scores.cpu().tolist()
+
+            for k, (box, label) in enumerate(zip(boxes, classes)):
+                if len(box) != 4:
+                    raise ValueError(
+                        f"Invalid input box of sample {i}, element {k} (expected 4 values, got {len(box)})"
+                    )
+                if type(label) != int:
+                    raise ValueError(
+                        f"Invalid input class of sample {i}, element {k}"
+                        f" (expected value of type integer, got type {type(label)})"
+                    )
+                annotation = {
+                    "id": annotation_id,
+                    "image_id": i,
+                    "bbox": box,
+                    "category_id": label,
+                    "area": box[2] * box[3],
+                    "iscrowd": 0,
+                }
+                if is_pred and scores is not None:
+                    score = scores[k]
+                    if type(score) != float:
+                        raise ValueError(
+                            f"Invalid input score of sample {i}, element {k}"
+                            f" (expected value of type float, got type {type(score)})"
+                        )
+                    annotation["score"] = score
+                annotations.append(annotation)
+                annotation_id += 1
+
+        classes = [{"id": i, "name": str(i)} for i in range(self.num_classes)]
```

Comment:
Apparently ``num_classes`` is required for mAP values per class, and if it's set to zero, the code should return only overall MAP and MAR values. However, if the "categories" defined here is empty, pycocotools will issue the warning

    Please run evaluate() first

and all the AP and AR values will be -1. It would be great if the user didn't have to know the number of classes. What if we set the (default) ``num_classes`` automatically:

    torch.cat(tuple(x["labels"] for x in preds + target)).unique()

Reply:
Well spotted, that is indeed a problem.

See the other thread why I'm unsure about removing the state variables.

What about using your suggested "unique classes workaround" only when `num_classes` is 0? Would that solve the problem?

Reply:
Yes, you could do that. Are you thinking that the "unique classes" value is not always correct, or in which case do you think the user would like to specify the number of classes? It may be that the user doesn't want the class-specific statistics, but is it enough to have a boolean switch?

Reply:
I cannot think of a case where classes should be used that are not included in `targets`, so I agree, a boolean switch should be enough.